flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/generator/CirculantGraph.java

```diff
@@ -0,0 +1,141 @@
+
+	private List<Long> signedOffsetList = new ArrayList<Long>();
+
+	/**
+	 * The {@link Graph} containing no edges.
+	 *
+	 * @param env the Flink execution environment
+	 * @param vertexCount number of vertices
+	 */
+	public CirculantGraph(ExecutionEnvironment env, long vertexCount, List<Long> offsetList) {
+		Preconditions.checkArgument(vertexCount >= MINIMUM_VERTEX_COUNT,
+			"Vertex count must be at least " + MINIMUM_VERTEX_COUNT);
+
+        if (offsetList != null && !offsetList.isEmpty()) {
+            Preconditions.checkArgument(new HashSet<>(offsetList).size() == offsetList.size(),
+                    "Offset must not be duplicated");
+
+            long maxOffset = vertexCount / 2;
+            for (long offset : offsetList) {
+                Preconditions.checkArgument(offset >= MINIMUM_OFFSET,
+                        "Offset must be at least " + MINIMUM_OFFSET);
+                Preconditions.checkArgument(offset <= maxOffset,
+                        "Offset must be at most " + maxOffset);
+
+                // add sign, ignore negative max offset when vertex count is even
+                signedOffsetList.add(offset);
+                if (!(vertexCount % 2 == 0 && offset == maxOffset)) {
+                    signedOffsetList.add(-offset);
+                }
+            }
+        }
+
+		this.env = env;
+		this.vertexCount = vertexCount;
+	}
+
+	@Override
+	public Graph<LongValue, NullValue, NullValue> generate() {
+		// Vertices
+		DataSet<Vertex<LongValue, NullValue>> vertices = GraphGeneratorUtils.vertexSequence(env, parallelism, vertexCount);
+
+        // Edges
+        LongValueSequenceIterator iterator = new LongValueSequenceIterator(0, this.vertexCount - 1);
+
+        DataSet<Edge<LongValue, NullValue>> edges = env
+                .fromParallelCollection(iterator, LongValue.class)
+                .setParallelism(parallelism)
+                .name("Edge iterators")
+                .flatMap(new LinkVertexToOffset(vertexCount, signedOffsetList))
+                .setParallelism(parallelism)
+                .name("Circulant graph edges");
+
+		// Graph
+		return Graph.fromDataSet(vertices, edges, env);
+	}
+
+    @FunctionAnnotation.ForwardedFields("*->f0")
+    public class LinkVertexToOffset
+            implements FlatMapFunction<LongValue, Edge<LongValue, NullValue>> {
+
+        private final long vertexCount;
+
+        private final List<Long> offsets;
+
+        private LongValue target = new LongValue();
+
+        private Edge<LongValue, NullValue> edge = new Edge<>(null, target, NullValue.getInstance());
+
+        public LinkVertexToOffset(long vertexCount, List<Long> offsets) {
+            this.vertexCount = vertexCount;
+            this.offsets = offsets;
+        }
+
+        @Override
+        public void flatMap(LongValue source, Collector<Edge<LongValue, NullValue>> out)
+                throws Exception {
+            // empty graph
+            if (offsets == null || offsets.isEmpty()) {
```

Comment:
I think better to not allow `offsets` to be `null`, then this block can be removed.